text-spacing: text-spacing-trim: implement trim-all for complex path

vitorroriz opened the pull request:
#### 85ed15ffe2c774c383a60e98b3087b73bd143d08
<pre>
text-spacing: text-spacing-trim: implement trim-all for complex path
<a href="https://bugs.webkit.org/show_bug.cgi?id=281761">https://bugs.webkit.org/show_bug.cgi?id=281761</a>
<a href="https://rdar.apple.com/137795771">rdar://137795771</a>

Reviewed by Sammy Gill.

Simple path (WidthIterator) already covers this behavior.
We should now do the same for the Complex path (ComplexTextController).

trim-all:
Set fullwidth opening punctuation, fullwidth closing punctuation, and fullwidth middle dot punctuation with half-width glyphs,
without regards for the position within the line nor for adjacent characters.

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::halfWidthFont const):

Canonical link: <a href="https://commits.webkit.org/286983@main">https://commits.webkit.org/286983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/010d8905f09ac7a0d86baaf0bd9e485cf26eed18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56270 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81800 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28517 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79352 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65418 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4566 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60505 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18552 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80302 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66290 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40805 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47878 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23788 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26840 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68988 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24123 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83223 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4615 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3113 "Found 1 new test failure: webrtc/vp8-then-h264.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68777 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4771 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68034 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17093 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12026 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10125 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4562 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7377 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4581 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8016 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6340 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -338,6 +340,17 @@ void ComplexTextController::collectComplexTextRuns()
+    auto getHalfWidthFontIfNeeded = [&] (RefPtr<const Font>& protectedHalfWidthFont, TextSpacing::CharactersData& charactersData) {
+        if (m_font.textSpacingTrim().shouldTrimSpacing(charactersData))
+            protectedHalfWidthFont = nextFont->halfWidthFont();
+        return protectedHalfWidthFont ? protectedHalfWidthFont : nextFont;
```
Doesn't it seem somewhat misleading for something called `getHalfWidthFontIfNeeded` to return something other than a pointer to a half width font? If we return `nextFont` does that represent some form of an error? Are we better off just returning a `nullptr` in this case and having the caller respond to that? Otherwise how can the caller distinguish between whether or not they got the half width font back?

sammygill commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -338,6 +340,17 @@ void ComplexTextController::collectComplexTextRuns()
     // would need to be updated accordingly too.
     nextFont = m_font.fontForCombiningCharacterSequence(baseOfString.first(currentIndex));
 
+    auto getHalfWidthFontIfNeeded = [&] (RefPtr<const Font>& protectedHalfWidthFont, TextSpacing::CharactersData& charactersData) {
```
Do we really need `protectedHalfWidthFont` and/or have it passed by reference here? Shouldn't whatever pointer we return here be responsible by the caller for protecting it (which we seem to do here already AFAICT)?

sammygill commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -378,6 +392,11 @@ void ComplexTextController::collectComplexTextRuns()
 
         nextFont = m_font.fontForCombiningCharacterSequence(baseOfString.subspan(previousIndex, currentIndex - previousIndex));
 
+        if (shouldProcessTextSpacingTrim && nextFont && !nextFont->isSystemFontFallbackPlaceholder()) {
```
We use this conditional twice which makes me think we could create some of helper function to refactor it into. That would probably also help us give a meaningful name for this logic to help improve the readability imo

sammygill commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -318,6 +318,7 @@ void ComplexTextController::collectComplexTextRuns()
     bool dontSynthesizeSmallCaps = !m_font.fontDescription().hasAutoFontSynthesisSmallCaps();
     bool engageAllSmallCapsProcessing = fontVariantCaps == FontVariantCaps::AllSmall || fontVariantCaps == FontVariantCaps::AllPetite;
     bool engageSmallCapsProcessing = engageAllSmallCapsProcessing || fontVariantCaps == FontVariantCaps::Small || fontVariantCaps == FontVariantCaps::Petite;
+    auto shouldProcessTextSpacingTrim = !m_font.textSpacingTrim().isSpaceAll();
```
Maybe bool for consistency? I don't feel strongly either way

sammygill commented on Source/WebCore/platform/text/TextSpacing.h:
```diff
@@ -151,4 +153,5 @@ inline WTF::TextStream& operator<<(WTF::TextStream& ts, const TextAutospace& val
     return ts;
 }
 
```
Accidental extra space?

sammygill commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -393,18 +393,12 @@ inline void WidthIterator::advanceInternal(TextIterator& textIterator, GlyphBuff
         TextSpacing::CharactersData charactersData = { .currentCharacter = character, .currentCharacterClass = TextSpacing::characterClass(character) };
-        halfWidthFont = getHalfWidthFontAndSetGlyphDataIfNeeded(glyphData, charactersData);
+        halfWidthFont = TextSpacing::getHalfWidthFontIfNeeded(*glyphData.font, fontDescription.textSpacingTrim(), charactersData);
+        glyphData.font = halfWidthFont ? halfWidthFont.get() : glyphData.font;
```
Maybe this should just be an `if` since the second result in the ternary is just the object itself?

sammygill commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -439,7 +433,8 @@ inline void WidthIterator::advanceInternal(TextIterator& textIterator, GlyphBuff
             TextSpacing::CharactersData charactersData = { .currentCharacter = character, .currentCharacterClass = TextSpacing::characterClass(character) };
-            halfWidthFont = getHalfWidthFontAndSetGlyphDataIfNeeded(glyphData, charactersData);
+            halfWidthFont = TextSpacing::getHalfWidthFontIfNeeded(*glyphData.font, fontDescription.textSpacingTrim(), charactersData);
+            glyphData.font = halfWidthFont ? halfWidthFont.get() : glyphData.font;
```
Ditto

sammygill commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -380,6 +389,12 @@ void ComplexTextController::collectComplexTextRuns()
 
         nextFont = m_font.fontForCombiningCharacterSequence(baseOfString.subspan(previousIndex, currentIndex - previousIndex));
 
+        if (shouldProcessTextSpacingTrim && nextFont && !nextFont->isSystemFontFallbackPlaceholder()) {
+            TextSpacing::CharactersData charactersData = { .currentCharacter = baseCharacter, .currentCharacterClass = TextSpacing::characterClass(baseCharacter) };
+            halfWidthFont = TextSpacing::getHalfWidthFontIfNeeded(*nextFont, m_font.textSpacingTrim(), charactersData);
+            nextFont = halfWidthFont ? halfWidthFont : nextFont;
+        }
```
Seems like we have some repition here. Is there an opportunity to refactor here?

sammygill replied:
(Just a minor observation and no concrete suggestions)

vitorroriz committed:
text-spacing: text-spacing-trim: implement trim-all for complex path
https://bugs.webkit.org/show_bug.cgi?id=281761
rdar://137795771

Reviewed by Sammy Gill.

Simple path (WidthIterator) already covers this behavior.
We should now do the same for the Complex path (ComplexTextController).

trim-all:
Set fullwidth opening punctuation, fullwidth closing punctuation, and fullwidth middle dot punctuation with half-width glyphs,
without regards for the position within the line nor for adjacent characters.

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::halfWidthFont const):

Canonical link: https://commits.webkit.org/286983@main
